contracts-bedrock/contracts/L1/ResourceMetering.sol

```diff
@@ -0,0 +1,185 @@
+        if (gasCost > usedGas) {
+            Burn.gas(gasCost - usedGas);
+        }
+    }
+
+    /**
+     * Unpacks resource params.
+     *
+     * @return prevBaseFee
+     * @return prevBoughtGas
+     * @return prevBlockNum
+     */
+    function getResourceParams()
+        public
+        view
+        returns (
+            uint128,
+            uint64,
+            uint64
+        )
+    {
+        return SlotPacking128x64x64.get(resources);
+    }
+
+    /**
+     * Updates resource parameters, only actually executes something once per block.
+     */
+    function updateResourceParams() public {
+        (uint128 prevBaseFee, uint64 prevBoughtGas, uint64 prevBlockNum) = getResourceParams();
+
+        uint256 blockDiff = block.number - prevBlockNum;
+        if (blockDiff == 0) {
+            return;
+        }
+
+        // Handle updating EIP-1559 style gas parameters. We use EIP-1559 to restrict the rate at
+        // which deposits can be created and therefore limit the potential for deposits to DoS the
+        // L2 system. Fee scheme is very similar to EIP-1559 with minor changes.
+        int256 gasUsedDelta = int256(uint256(prevBoughtGas)) - TARGET_RESOURCE_LIMIT;
+        int256 baseFeeDelta = (int256(uint256(prevBaseFee)) * gasUsedDelta) /
+            TARGET_RESOURCE_LIMIT /
+            BASE_FEE_MAX_CHANGE_DENOMINATOR;
+
+        // Clamp the value of the base fee between the minimum and maximum values.
+        uint128 newBaseFee = uint128(
+            uint256(
+                SignedMath.min(
+                    SignedMath.max(
+                        int256(uint256(prevBaseFee)) + baseFeeDelta,
+                        int256(MINIMUM_BASE_FEE)
+                    ),
+                    int256(uint256(type(uint128).max))
+                )
+            )
+        );
+
+        // If we skipped more than one block, we also need to account for every empty block. Empty
+        // block means there was no demand for deposits in that block, so we should reflect this
+        // lack of demand in the fee.
+        if (blockDiff > 1) {
+            // Repeatedly reduce newBaseFee by 7/8 of its value based on the number of skipped
+            // blocks. We break this math into chunks of x=40 (at a maximum) to avoid an overflow.
+            // TODO: Replace with solmate's powWad when v7 is released, it's close to constant gas
+            // for any value of n.
+            uint256 n = blockDiff - 1;
+            uint256 tempBaseFee = uint256(newBaseFee);
+            for (uint256 i = 0; i < n; i += 40) {
+                // Execution is bounded by the fact that we'll eventually get down below the
+                // minimum base fee. We can stop early to avoid wasting gas.
+                if (tempBaseFee <= uint256(MINIMUM_BASE_FEE)) {
+                    break;
+                }
+
+                uint256 x = Math.min(n - i, 40);
+                tempBaseFee = (tempBaseFee * (7**x)) / (8**x);
```

Comment:
I'm happy this doesn't overflow.
`8^42 < 2^128 -1 < 8 ^ 43` => 7^40 and 8^40 fit within 128 bits.
The basefee starts as a 128 bit number, and then two 128 bit numbers multiplied fit in a 256 bit number. We then immediately divide `basefee * 7^n` by `8^n`. As `8^n > 7^n`, the resulting variable fits back into 128 bits.
Then we proceed with the loop with the same invariant (basefee fits in 128 bit number). See above for loop termination.